blender: use correct Python version

amarshall opened the pull request:
## Description of changes

See individual commit messages for elaboration.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

amarshall committed:
blender: use correct Python version

This partially reverts commit 46861969d7e56eb98451d8a270b160c7cbbb37ea.
Using nixpkgs’s default Python was previously necessary to workaround a
problem that no longer exists: materialx always using the top-level
python3 instead of that which matches its dependents.

Using the same Python version as the upstream Blender release is desired
to ensure compatibility not only Blender itself but with the addon
ecosystem.

Also move the selection of non-nixpkgs-default Python to all-packages
rather than the pkg’s function args as that is the current preference
(as documented in
./pkgs/by-name/README.md#changing-implicit-attribute-defaults).

dotlambda commented on pkgs/top-level/aliases.nix:
```diff
@@ -865,6 +865,7 @@ mapAliases ({
   markmind = throw "markmind has been removed from nixpkgs, because it depended on an old version of electron"; # Added 2023-09-12
   marwaita-manjaro = lib.warn "marwaita-manjaro has been renamed to marwaita-darkcyan" marwaita-darkcyan; # Added 2024-07-01
   marwaita-peppermint = lib.warn "marwaita-peppermint has been renamed to marwaita-red" marwaita-red; # Added 2024-07-01
+  materialx = python3Packages.materialx; # Added 2024-07-09
```
This should be in `all-packages.nix`.

dotlambda commented on pkgs/applications/misc/blender/default.nix:
```diff
@@ -71,7 +71,7 @@
   potrace,
   pugixml,
-  python3Packages, # must use instead of python3.pkgs, see https://github.com/NixOS/nixpkgs/issues/211340
+  python311Packages, # must use instead of python3.pkgs, see https://github.com/NixOS/nixpkgs/issues/211340
```
So Blender is not compatible with Python 3.12?

gador replied:
Blender runs fine on `python3.12` This was changed in #325292, too. If we don't follow `python3Packages` we introduce errors between the python versions (The ones actually fixed by that PR)

dotlambda replied:
> Blender runs fine on `python3.12`

As pointed out, that doesn't mean addons do.

gador replied:
See the comment by @kirillrdy (https://github.com/NixOS/nixpkgs/pull/325292#discussion_r1667893187)

dotlambda replied:
> If we don't follow `python3Packages` we introduce errors between the python versions

Not once this PR or https://github.com/NixOS/nixpkgs/pull/326041 is merged.

dotlambda replied:
> See the comment by @kirillrdy ([#325292 (comment)](https://github.com/NixOS/nixpkgs/pull/325292#discussion_r1667893187))

That's just personal preference regarding style and has nothing to do with whether Python 3.11 or 3.12 should be used.

gador replied:
I had to update the previous `python311Packages` to `python312Packages` in order to not have a version mismatch. The changes to `python3Packages` were not only cosmetic, but also functional, because we don't have to change the version by hand after each python upgrade. 
I agree, once #326041 is merged, this shouldn't matter :laughing:

kirillrdy replied:
@dotlambda If I understand correctly, blender probably should be using python 3.11 due to some plugins still not compatible with 3.12. As for recommendation to change default versions in all-packages.nix or python-packages.nix, I see it as official recommendation rather than person preference, please correct me otherwise.

dotlambda replied:
> I see it as official recommendation rather than person preference

Reference?

kirillrdy replied:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md#changing-implicit-attribute-defaults

SuperSandro2000 commented:
Please rebase on master

kirillrdy commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29582,6 +29582,7 @@ with pkgs;
 
   blender = callPackage  ../applications/misc/blender {
     openexr = openexr_3;
+    python3Packages = python311Packages;
```
maybe we should add a comment, what to test ? and/or what breaks

amarshall committed:
blender: verify expected Python version during build

Upstream builds and tests with a specific Python version. Further, the
Blender addon ecosystem generally expects that version to be used, so
ensure that the correct version is used when upstream changes. Python
often makes backwards-incompatible changes in “minor” releases.

To quote upstream source code:

> Blender only supports a single Python version at the moment.

Perform the check during preConfigure since we can check it then and so
it avoids wasting time configuring+building and then failing.

gador commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 326044` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>blender</li>
<li>blender-hip</li>
</ul>
</details>

dotlambda commented:
> `python3{11,12}Packages.numpy_1` are both broken on staging

fixed by https://github.com/NixOS/nixpkgs/pull/435176
